cmd/kubelet/app/server.go

```diff
@@ -946,28 +946,29 @@ func RunKubelet(kubeFlags *options.KubeletFlags, kubeCfg *kubeletconfiginternal.
 
 func startKubelet(k kubelet.Bootstrap, podCfg *config.PodConfig, kubeCfg *kubeletconfiginternal.KubeletConfiguration, kubeDeps *kubelet.Dependencies, enableServer bool) {
 	wg := sync.WaitGroup{}
+	once := sync.Once{}
 
 	// start the kubelet
 	wg.Add(1)
 	go wait.Until(func() {
-		wg.Done()
+		once.Do(func() { wg.Done() })
 		k.Run(podCfg.Updates())
 	}, 0, wait.NeverStop)
 
 	// start the kubelet server
 	if enableServer {
 		wg.Add(1)
-		go wait.Until(func() {
+		go func() {
 			wg.Done()
 			k.ListenAndServe(net.ParseIP(kubeCfg.Address), uint(kubeCfg.Port), kubeDeps.TLSOptions, kubeDeps.Auth, kubeCfg.EnableDebuggingHandlers, kubeCfg.EnableContentionProfiling)
-		}, 0, wait.NeverStop)
+		}()
 	}
 	if kubeCfg.ReadOnlyPort > 0 {
 		wg.Add(1)
-		go wait.Until(func() {
+		go func() {
 			wg.Done()
 			k.ListenAndServeReadOnly(net.ParseIP(kubeCfg.Address), uint(kubeCfg.ReadOnlyPort))
-		}, 0, wait.NeverStop)
+		}()
 	}
 	wg.Wait()
```

Comment:
What is the point of this wait group? It looks like it's waiting for the goroutines to start, but why? It's not actually checking that the servers in those go routines have started...

Reply:
If listen and serve failed , process will exit.
```
// ListenAndServeKubeletReadOnlyServer initializes a server to respond to HTTP network requests on the Kubelet.
func ListenAndServeKubeletReadOnlyServer(host HostInterface, resourceAnalyzer stats.ResourceAnalyzer, address net.IP, port uint, runtime kubecontainer.Runtime) {
	glog.V(1).Infof("Starting to listen read-only on %s:%d", address, port)
	s := NewServer(host, resourceAnalyzer, nil, false, false, runtime, nil)

	server := &http.Server{
		Addr:           net.JoinHostPort(address.String(), strconv.FormatUint(uint64(port), 10)),
		Handler:        &s,
		MaxHeaderBytes: 1 << 20,
	}
	glog.Fatal(server.ListenAndServe())
}
```

Reply:
I don't think that answers my question. Why can't you just remove the waitgroup entirely? It doesn't appear to be doing anything useful here.

Reply:
The wa.Add（） after wg.Wait may cause a panic.

Reply:
And I thought wg wait for server really startup.

Reply:
I should remove the waitgroup.